### PR TITLE
[Employee (Admin cabinet)] Add activate employee

### LINKED ISSUE #6148: [Employee (Admin cabinet)]There is no button 'Deactivate an employee' to deactivate an active employee card
Environment: Windows 10 Pro 20H2
Google Chrome 113.0.5672.93
Reproducible: always
Build found: 09.08.2023
Preconditions

1.  The user should be logged in https://www.greencity.social/#/greenCity as a Super Admin
2. In the tab 'Employees' created an employee's card with the status 'Active'

Steps to reproduce

1. Click on the 'Employee' tab
2. In the tab 'Employee' choose an employee's card with the status 'Active'
3. Click on three dots from the right in the employee card row

Actual result

Appears pop-up window with the icon 'Delete employee'

Expected result

Appears pop-up window with the icon 'Deactivate an employee'

Screenshot_3

User story https://github.com/ita-social-projects/GreenCity/issues/4617

"Bug", Priority ("Medium"), Severity ("Major"), Type ("UI"), "API" (for back-end bugs)

### PR DESCRIPTION
**Pull request by ospodaryk**
# GreenCityUBS PR
[[Employee (Admin cabinet)]There is no button 'Deactivate an employee' to deactivate an active employee card #6148
](https://github.com/ita-social-projects/GreenCity/issues/6148)

## Summary Of Changes :fire:
Now we have methods to activate user and employee

## Added
Added methods to activate user and employee and tests to it.

# PR Checklist Forms

_(to be filled out by PR submitter)_
- [ ] Code is up-to-date with the `dev` branch.
- [x] You've successfully built and run the tests locally.
- [x] There are new or updated unit tests validating the change.
- [x] JIRA/ Github Issue number & title in PR title (ISSUE-XXXX: Ticket title)
- [x] This template filled (above this section).
- [x] Sonar's report does not contain bugs, vulnerabilities, security issues, code smells ar duplication
- [x] `NEED_REVIEW` and `READY_FOR_REVIEW` labels are added.
- [x] All files reviewed before sending to reviewers

### CODE REVIEW

File: core/src/main/java/greencity/configuration/SecurityConfig.java [hunk 1]
```diff
@@ -157,6 +157,10 @@ protected void configure(HttpSecurity http) throws Exception {
                 SUPER_ADMIN_LINK + "/deactivateCourier/{id}",
                 SUPER_ADMIN_LINK + "/switchTariffStatus/{tariffId}")
             .hasAnyRole(ADMIN, UBS_EMPLOYEE)
+            .antMatchers(HttpMethod.PATCH,
+                SUPER_ADMIN_LINK + "/activateCourier/{id}",
+                SUPER_ADMIN_LINK + "/switchTariffStatus/{tariffId}")
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/configuration/SecurityConfig.java]: duplication

File: core/src/main/java/greencity/configuration/SecurityConfig.java [hunk 2]
```diff
@@ -157,6 +157,10 @@ protected void configure(HttpSecurity http) throws Exception {
                 SUPER_ADMIN_LINK + "/deactivateCourier/{id}",
                 SUPER_ADMIN_LINK + "/switchTariffStatus/{tariffId}")
             .hasAnyRole(ADMIN, UBS_EMPLOYEE)
+            .antMatchers(HttpMethod.PATCH,
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/configuration/SecurityConfig.java]: why did you add one more `.antMatchers`?

File: core/src/main/java/greencity/controller/ManagementEmployeeController.java [hunk 3]
```diff
@@ -138,6 +138,27 @@ public ResponseEntity<HttpStatus> deleteEmployee(@PathVariable Long id) {
...
+     */
+    @ApiOperation(value = "Activate employee")
+    @ApiResponses(value = {
+        @ApiResponse(code = 200, message = HttpStatuses.OK),
+        @ApiResponse(code = 400, message = HttpStatuses.BAD_REQUEST),
+        @ApiResponse(code = 401, message = HttpStatuses.UNAUTHORIZED),
+        @ApiResponse(code = 403, message = HttpStatuses.FORBIDDEN),
+        @ApiResponse(code = 404, message = HttpStatuses.NOT_FOUND)
+    })
+    @PreAuthorize("@preAuthorizer.hasAuthority('DEACTIVATE_EMPLOYEE', authentication)")
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/ManagementEmployeeController.java:155]: why do we use `DEACTIVATE_EMPLOYEE` authority for activation?

File: core/src/main/java/greencity/controller/ManagementEmployeeController.java [hunk 4]
```diff
@@ -138,6 +138,27 @@ public ResponseEntity<HttpStatus> deleteEmployee(@PathVariable Long id) {
...
+    @PutMapping("/activate-employee/{id}")
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/ManagementEmployeeController.java:156]: please add security configs for this endpoint

File: core/src/main/java/greencity/controller/ManagementEmployeeController.java (hunk 3 above)
Comment (ospodaryk) [reply core/src/main/java/greencity/controller/ManagementEmployeeController.java:155]: Because if user have this authority he also can ACTIVATE_EMPLOYEE

File: core/src/main/java/greencity/controller/SuperAdminController.java [hunk 5]
```diff
@@ -462,6 +462,23 @@ public ResponseEntity<CourierDto> deactivateCourier(@PathVariable Long id) {
...
+     * @param id - courier id that will need to be deleted;
+     */
+    @ApiOperation(value = "Activate courier's by Id")
+    @ApiResponses(value = {
+        @ApiResponse(code = 200, message = HttpStatuses.OK),
+        @ApiResponse(code = 400, message = HttpStatuses.BAD_REQUEST),
+        @ApiResponse(code = 401, message = HttpStatuses.UNAUTHORIZED),
+        @ApiResponse(code = 403, message = HttpStatuses.FORBIDDEN)
+    })
+    @PatchMapping("/activateCourier/{id}")
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/SuperAdminController.java]: Could we have 404 here if the courier with the requested id does not exist?

File: core/src/main/java/greencity/configuration/SecurityConfig.java (hunk 1 above)
Comment (ospodaryk) [reply core/src/main/java/greencity/configuration/SecurityConfig.java]: fixed

File: core/src/main/java/greencity/configuration/SecurityConfig.java (hunk 2 above)
Comment (ospodaryk) [reply core/src/main/java/greencity/configuration/SecurityConfig.java]: removed

File: service-api/src/main/java/greencity/constant/ErrorMessage.java [hunk 6]
```diff
@@ -137,6 +138,8 @@ public final class ErrorMessage {
     public static final String MIN_MAX_VALUE_RESTRICTION = "Min and Max fields must have different values";
     public static final String EMPLOYEE_WITH_CURRENT_UUID_WAS_NOT_DEACTIVATED = "Employee with current uuid was not "
         + "deactivated.";
+    public static final String EMPLOYEE_WITH_CURRENT_UUID_WAS_NOT_ACTIVATED = "Employee with current uuid was not "
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/constant/ErrorMessage.java]: please add `uuid` to the log

File: service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java [hunk 7]
```diff
@@ -475,6 +475,19 @@ public CourierDto deactivateCourier(Long id) {
         return modelMapper.map(courier, CourierDto.class);
     }
 
+    @Override
+    @Transactional
+    public CourierDto activateCourier(Long id) {
+        Courier courier = courierRepository.findById(id).orElseThrow(
+            () -> new NotFoundException(ErrorMessage.COURIER_IS_NOT_FOUND_BY_ID + id));
+        if (CourierStatus.ACTIVE == courier.getCourierStatus()) {
+            throw new BadRequestException(ErrorMessage.CANNOT_ACTIVATE_COURIER + courier.getId());
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java]: please add more informative message here

File: service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java [hunk 8]
```diff
@@ -475,6 +475,19 @@ public CourierDto deactivateCourier(Long id) {
...
+        }
+        deactivateTariffsForChosenParamRepository.deactivateTariffsByCourier(id);
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java]: why should we deactivate all tariffs before activating the courier?

File: service/src/main/java/greencity/service/ubs/UBSManagementEmployeeServiceImpl.java [hunk 9]
```diff
@@ -289,6 +289,25 @@ public void deactivateEmployee(Long id) {
...
 
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    @Transactional
+    public void activateEmployee(Long id) {
+        Employee employee = employeeRepository.findById(id)
+            .orElseThrow(() -> new NotFoundException(ErrorMessage.EMPLOYEE_NOT_FOUND + id));
+        if (employee.getEmployeeStatus().equals(EmployeeStatus.INACTIVE)) {
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSManagementEmployeeServiceImpl.java]: we could use `== `for enums

File: service/src/main/java/greencity/service/ubs/UBSManagementEmployeeServiceImpl.java [hunk 10]
```diff
@@ -289,6 +289,25 @@ public void deactivateEmployee(Long id) {
...
+            employee.setEmployeeStatus(EmployeeStatus.ACTIVE);
+            try {
+                userRemoteClient.activateEmployee(employee.getUuid());
+            } catch (HystrixRuntimeException e) {
+                throw new BadRequestException("Employee with current uuid doesn't exist: " + employee.getUuid());
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSManagementEmployeeServiceImpl.java]: Do we have `HystrixRuntimeException`  only if `Employee with current uuid doesn't exist`?

File: core/src/main/java/greencity/controller/SuperAdminController.java [hunk 11]
```diff
@@ -462,6 +462,23 @@ public ResponseEntity<CourierDto> deactivateCourier(@PathVariable Long id) {
...
+    public ResponseEntity<CourierDto> activateCourier(@PathVariable Long id) {
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/SuperAdminController.java]: Why do we have the different responses for courier and employee activation?
We retund CourierDto for courier activation and only the response code for employee activation.

File: core/src/main/java/greencity/controller/ManagementEmployeeController.java (hunk 4 above)
Comment (ospodaryk) [reply core/src/main/java/greencity/controller/ManagementEmployeeController.java:156]: added

File: core/src/main/java/greencity/controller/SuperAdminController.java (hunk 5 above)
Comment (ospodaryk) [reply core/src/main/java/greencity/controller/SuperAdminController.java]: yes, added

File: service/src/main/java/greencity/service/ubs/UBSManagementEmployeeServiceImpl.java (hunk 10 above)
Comment (ospodaryk) [reply service/src/main/java/greencity/service/ubs/UBSManagementEmployeeServiceImpl.java]: yes

File: service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java (hunk 8 above)
Comment (ospodaryk) [reply service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java]: fixed

File: service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java (hunk 7 above)
Comment (ospodaryk) [reply service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java]: fixed

File: service-api/src/main/java/greencity/constant/ErrorMessage.java (hunk 6 above)
Comment (ospodaryk) [reply service-api/src/main/java/greencity/constant/ErrorMessage.java]: added

File: core/src/main/java/greencity/configuration/SecurityConfig.java [hunk 12]
```diff
@@ -155,6 +156,7 @@ protected void configure(HttpSecurity http) throws Exception {
             .hasAnyRole(ADMIN, UBS_EMPLOYEE)
             .antMatchers(HttpMethod.PATCH,
                 SUPER_ADMIN_LINK + "/deactivateCourier/{id}",
+                SUPER_ADMIN_LINK + "/activateCourier/{id}",
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/configuration/SecurityConfig.java]: never used

File: core/src/test/java/greencity/controller/SuperAdminControllerTest.java [hunk 13]
```diff
@@ -1039,4 +1039,5 @@ void deactivateCourier() throws Exception {
         mockMvc.perform(patch(ubsLink + "/deactivateCourier/{id}", 1L)).andExpect(status().isOk());
         verify(superAdminService).deactivateCourier(1L);
     }
+
```
Comment (RomaMocherniuk) [inline core/src/test/java/greencity/controller/SuperAdminControllerTest.java]: do we need these changes?

File: service-api/src/main/java/greencity/constant/ErrorMessage.java [hunk 14]
```diff
@@ -84,6 +84,7 @@ public final class ErrorMessage {
     public static final String LOCATION_IS_DEACTIVATED_FOR_TARIFF = "Location is deactivated for tariff: ";
     public static final String COURIER_IS_NOT_FOUND_BY_ID = "Couldn't found courier by id: ";
     public static final String CANNOT_DEACTIVATE_COURIER = "Courier is already deactivated with id: ";
+    public static final String CANNOT_ACTIVATE_COURIER_ALREADY_ACTIVATED = "Courier is already activated with id: ";
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/constant/ErrorMessage.java]: never used

File: service/src/main/java/greencity/service/ubs/UBSManagementEmployeeServiceImpl.java [hunk 15]
```diff
@@ -289,6 +289,25 @@ public void deactivateEmployee(Long id) {
...
+    @Transactional
+    public void activateEmployee(Long id) {
+        Employee employee = employeeRepository.findById(id)
+            .orElseThrow(() -> new NotFoundException(ErrorMessage.EMPLOYEE_NOT_FOUND + id));
+        if (employee.getEmployeeStatus() == EmployeeStatus.INACTIVE) {
+            employee.setEmployeeStatus(EmployeeStatus.ACTIVE);
+            try {
+                userRemoteClient.activateEmployee(employee.getUuid());
+            } catch (HystrixRuntimeException e) {
+                throw new BadRequestException("Employee with current uuid doesn't exist: " + employee.getUuid());
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSManagementEmployeeServiceImpl.java]: Could we move it to constant?

**Review by sergoliarnik — APPROVED** [review]

**Review by Yurii-Midianyi — APPROVED** [review]

**Review by ABbondar — APPROVED** [review]

**Review by ABbondar — APPROVED** [review]

**Review by RomaMocherniuk — APPROVED** [review]
